src/resources/JsResolver.ts

```diff
@@ -18,8 +19,32 @@ export class JsResolver {
       );
     }
 
-    const requestTemplateContent = fs.readFileSync(this.config.path, 'utf8');
-    return this.processTemplateSubstitutions(requestTemplateContent);
+    // process with esbuild
+    // this will:
+    // - Bundle the code into one file if necessary
+    // - Transpile typescript to javascript if necessary
+
+    const buildResult = buildSync({
+      entryPoints: [this.config.path],
+      bundle: true,
+      write: false,
+      external: ['@aws-appsync/utils'],
+      format: 'esm',
```

Comment:
@bboure Thanks for sharing. Love it! I didn't know getting TS support + bundling was that simple. 🚀

I'm now using something similar in one of my CDK projects. However, I found one case where the resulting JS code didn't work. Here's a simplified version of the resolver:

```ts
import { util } from '@aws-appsync/utils'
import type { Context } from '@aws-appsync/utils'

export function response(ctx: Context) {
  const { result, error } = ctx

  if (error) {
    return util.error(error.message, error.type)
  }
  if (result.items.length === 0) {
    return util.error('user not found')
  }

  const user = result.items[0]

  return {
    id: user.UserID,
    slack: {
      enabled: user.Slack?.Enabled ?? false,
      webhookUrl: user.Slack?.WebhookURL,
      channel: user.Slack?.Channel,
    },
  }
}
```

Transpiled to JS with esbuild:

```js
// esbuild --bundle --external:@aws-appsync/utils --format=esm resolvers/getUser.ts
import { util } from "@aws-appsync/utils";
function response(ctx) {
  var _a, _b, _c, _d;
  const { result, error } = ctx;
  if (error) {
    return util.error(error.message, error.type);
  }
  if (result.items.length === 0) {
    return util.error("user not found");
  }
  const user = result.items[0];
  return {
    id: user.UserID,
    slack: {
      enabled: (_b = (_a = user.Slack) == null ? void 0 : _a.Enabled) != null ? _b : false,
      webhookUrl: (_c = user.Slack) == null ? void 0 : _c.WebhookURL,
      channel: (_d = user.Slack) == null ? void 0 : _d.Channel
    }
  };
}
export {
  response
};
```

However, at runtime, the resolver will return `ReferenceError: _a is not defined`.

I don't know why that is tbh. The fix was to target ES2020, which introduced optional chaining:

```js
// esbuild --bundle --external:@aws-appsync/utils --format=esm --target=es2020 resolvers/getUser.ts
import { util } from "@aws-appsync/utils";
function response(ctx) {
  const { result, error } = ctx;
  if (error) {
    return util.error(error.message, error.type);
  }
  if (result.items.length === 0) {
    return util.error("user not found");
  }
  const user = result.items[0];
  return {
    id: user.UserID,
    slack: {
      enabled: user.Slack?.Enabled ?? false,
      webhookUrl: user.Slack?.WebhookURL,
      channel: user.Slack?.Channel
    }
  };
}
export {
  response
};
```

This resolver works just fine.

So you might consider passing `target: 'es2020'` too.

Reply:
Thanks!
That's super useful! I definitely need to investigate more and do more testing.
In my research, I noticed I needed at least to target es2018. For some reason, I omitted this here.